### PR TITLE
[vnet][0] setup virtual netstack

### PR DESCRIPTION
**Pull request by nklaassen**
This is the first in a series of PRs implementing Teleport VNet [RFD](https://github.com/gravitational/teleport/blob/rfd/0163-vnet/rfd/0163-vnet.md). [child](https://github.com/gravitational/teleport/pull/40893)

This commit sets up virtual networking based on the gvisor netstack, and implements a unit testing strategy for exercising VNet's virtual network without relying on a real TUN interface from the host OS.

### CODE REVIEW

**Review by avatus — APPROVED** [review]

File: lib/vnet/vnet.go
Comment (ibeckermayer) [reply lib/vnet/vnet.go:235]: Looks good!

File: lib/vnet/vnet.go
Comment (ibeckermayer) [reply lib/vnet/vnet.go:141]: I took a stab at what I had in mind here: https://github.com/gravitational/teleport/compare/nklaassen/vnet0...isaiah/vnet0-suggestion?expand=1

The basic idea was to try and make the flow of a `tcp.ForwardRequest` through the system more readily apparent via the types. It also incidentally eliminated the need for anything related to our discussion [here](https://github.com/gravitational/teleport/pull/40889#discussion_r1581733072). 

Your call whether to take it or leave it.

File: lib/vnet/vnet_test.go
Comment (ibeckermayer) [inline lib/vnet/vnet_test.go:206]: My editor complains `error var fakeTUNClosedErr should have name of the form errFoo (ST1012)go-staticcheck`

File: lib/vnet/vnet_test.go
Comment (ibeckermayer) [inline lib/vnet/vnet_test.go:62]: nit: this comment makes it sound like a test-specific thing, but in fact this is exactly what we're using in production

File: lib/vnet/vnet_test.go
Comment (ibeckermayer) [inline lib/vnet/vnet_test.go:171]: It seems like this `\n` is load bearing, but why? (Maybe worth a comment).

**Review by ibeckermayer — APPROVED** [review]
Thanks for bearing with all my nits, I just find this project intrinsically interesting and am doing my best to understand the inner workings.

See https://github.com/gravitational/teleport/pull/40889#discussion_r1583602208 before merging

File: lib/vnet/vnet.go
Comment (nklaassen) [reply lib/vnet/vnet.go:141]: I do appreciate the detailed review, and I'm not ignoring this comment, I'm just working out the details of this API at the top of the PR stack where I start to see the real implementation, coming soon ™️

Comment (nklaassen) [reply lib/vnet/vnet.go:141]: I'm iterating on this in https://github.com/gravitational/teleport/pull/41031

There's a reason for doing the thing where the handler has to call a function to get the `net.Conn`:
```
// TCPHandler defines the behavior for handling TCP connections from VNet.
//
// Implementations should attempt to dial the target application and return any errors before calling
// [connector] to complete the TCP handshake and get the TCP conn. This is so that clients will see that the
// TCP connection was refused, instead of seeing a successful TCP dial that is immediately closed.
type TCPHandler interface {
	HandleTCP(ctx context.Context, connector func() (net.Conn, error)) error
}
```

File: lib/vnet/vnet_test.go
Comment (nklaassen) [reply lib/vnet/vnet_test.go:175]: so the bufio scanner can read a line, updated the comment. this test is mostly being replaced later otherwise i'd remove the bufio

File: lib/vnet/vnet_test.go
Comment (nklaassen) [reply lib/vnet/vnet_test.go:62]: it is a test-specific thing, update the comment:

```
// Create an isolated userspace networking stack that can be manipulated from test code. It will be
// connected to the VNet over the TUN interface. This emulates the host networking stack.
// This is a completely separate gvisor network stack than the one that will be created in NewManager -
// the two will be connected over a fake TUN interface. This exists so that the test can setup IP routes
// without affecting the host running the Test.
```

**Pull request merged**
